src/smt/theory_str_noodler/theory_str_noodler_model.cpp

```diff
@@ -74,16 +79,34 @@ namespace smt::noodler {
     
         app * mk_value(model_generator & m, expr_ref_vector const & values) override {
             if (!length_relevant) {
-                // because the length is not relevant, we can return anything, so we return empty string
+                // because the length is not relevant, we can return any string; let the
+                // standard sequence factory pick a fresh value when available, to avoid
+                // collapsing the string universe to a single element.
+                // try to obtain the seq_factory from the proto_model owned by
+                // the model generator and use it to pick a fresh string value
+                // when available.
+                proto_model& mdl = m.get_model();
+                value_factory* vf = mdl.get_factory(m_util_s.get_family_id());
+                if (auto* f = dynamic_cast<seq_factory*>(vf)) {
+                    expr* v = f->get_fresh_value(m_util_s.str.mk_string_sort());
+                    return to_app(v);
+                }
                 return m_util_s.str.mk_string(zstring());
             } else {
                 // values[0] contain the length of str_var, so we return some string of this length
                 bool is_int;
                 rational val(0);
                 SASSERT(values.size() == 1);
                 VERIFY(m_util_a.is_numeral(values[0], val, is_int) && is_int);
-                std::vector<unsigned> res(val.get_unsigned(), 'a'); // we can return anything, so we will just fill it with 'a'
-                return m_util_s.str.mk_string(zstring(res.size(), res.data()));
+                unsigned len = val.get_unsigned();
+                std::vector<unsigned> res(len, 'a');
+                expr* v = m_util_s.str.mk_string(zstring(res.size(), res.data()));
+                proto_model& mdl = m.get_model();
+                value_factory* vf = mdl.get_factory(m_util_s.get_family_id());
+                if (auto* f = dynamic_cast<seq_factory*>(vf)) {
+                    f->register_value(v);
+                }
+                return to_app(v);
```

Comment:
Shouldn't we return a fresh value also here? If we get some length variable both with lengths 2, they will have "aa" as model.